subprojects/build-init/src/test/groovy/org/gradle/api/tasks/wrapper/WrapperTest.groovy

```diff
@@ -214,4 +214,30 @@ class WrapperTest extends AbstractTaskTest {
         cleanup:
         Locale.setDefault(originalLocale)
     }
+
+    def "distributionSha256Sum declared in gradle-wrapper.properties is preserved by wrapper task"() {
+        given:
+        wrapper.setDistributionSha256Sum(null)
+        expectedTargetWrapperProperties << "distributionSha256Sum = somehash"
+
+        when:
+        execute(wrapper)
+
+        then:
+        def properties = GUtil.loadProperties(expectedTargetWrapperProperties)
+        properties.getProperty("distributionSha256Sum") == "somehash"
+    }
```

Comment:
I think this isn't quite right.

Consider what happens when someone upgrades the wrapper without the checksum:
```
gradle wrapper --gradle-version=7.5 --gradle-distribution-sha256-sum=cb87f222c5585bd46838ad4db78463a5c5f3d336e5e2b98dc7c0c586527351c2

gradle wrapper --gradle-version=7.5.1

gradle help <-- fails on checksum verification
```

I think this is not great behavior for a couple of reasons:
1. Users are now broken and the only way to get back to a working state is to revert the changes to `gradle-wrapper.properties` or manually edit the file to include the correct checksum. There's no way to fix this through Gradle.
2. Users are going to see that their distribution has failed verification immediately after upgrading. This is going to either unnecessarily alarm someone or they'll get used to see an "expected" error after updating.

What if instead of preserving the checksum in the file blindly, we fail to update the wrapper script if the URL to the distribution changes, the properties file has a checksum and the user does not specify a new checksum? I think changing the wrapper URL and keeping the old checksum will always be wrong.

```
gradle wrapper --gradle-version=7.5 --gradle-distribution-sha256-sum=cb87f222c5585bd46838ad4db78463a5c5f3d336e5e2b98dc7c0c586527351c2

gradle wrapper --gradle-version=7.5.1 <-- fail here that we cannot update the wrapper because the wrapper has a checksum and the user didn't provide a new one
```

Reply:
You are right, we missed this usecase.  The reason being is that there are two different cases with regards to wrapper updates: running the wrapper task, and editing the `gradle.properties` files directly. We were focusing on the second one only.

I think those two cases need different changes in the implementation.

Your proposal make sense to me, although I think failing the build is a bit too harsh in this case. We should only emit a warning notifying the user that the checksum is no longer present in the `gradle.properties` file. 

I also think that we should treat editing `gradle.properties` file directly as a valid usecase. A lot of people (me included) upgrades their Gradle build by changing the URL (and the checksum) and not even running the `wrapper` task. This is particularly useful for larger builds with longer configuration times. In this case, preserving the checksum makes sense. This PR should be adjusted such that the checksum is kept only if `GradleVersion.current() == task['wrapper'].gradleVersion`.

WDYT?